File: internal/service/flow_definition_service.go
package service

import (
	"encoding/json"
	"errors"
	"fmt"
	"math/rand"
	"time"

	"github.com/nexus-flow/nexus/internal/model"
	"github.com/nexus-flow/nexus/internal/repository"
)

type FlowDefinitionInfoDTO struct {
	ID              int64       `json:"id"`
	FlowKey         string      `json:"flowKey"`
	FlowName        string      `json:"flowName"`
	FlowType        string      `json:"flowType"`
	FlowContent     string      `json:"flowContent"`
	Remark          string      `json:"remark,omitempty"`
	CreatedAt       *time.Time  `json:"createdAt,omitempty"`
	FlowInputParams interface{} `json:"flowInputParams,omitempty"`
	FlowOutputParams interface{} `json:"flowOutputParams,omitempty"`
	FlowVariables   interface{} `json:"flowVariables,omitempty"`
}

const defaultFlowContent = `[{"key":"start_1","name":"开始","outgoings":["end_1"],"incomings":[],"elementType":"START"},{"key":"end_1","name":"结束","outgoings":[],"incomings":["start_1"],"elementType":"END"}]`

func CreateFlowDefinition(flowName, flowType, remark string, inputParams, outputParams interface{}) error {
	flowKey := generateFlowKey()
	fd := &model.FlowDefinition{
		FlowKey:     flowKey,
		FlowName:    flowName,
		FlowType:    flowType,
		FlowContent: defaultFlowContent,
		Remark:      remark,
	}
	if err := repository.CreateFlowDefinition(fd); err != nil {
		return err
	}

	saveFlowDefinitionParams(fd.ID, inputParams, outputParams)
	return nil
}

func UpdateFlowDefinition(id int64, flowName, flowType, remark string, inputParams, outputParams interface{}) error {
	fd := &model.FlowDefinition{
		FlowName: flowName,
		FlowType: flowType,
		Remark:   remark,
	}
	fd.ID = id
	if err := repository.UpdateFlowDefinition(fd); err != nil {
		return err
	}

	saveFlowDefinitionParams(id, inputParams, outputParams)
	return nil
}

func SaveFlowDefinitionContent(id int64, content string, variables interface{}) error {
	if err := repository.UpdateFlowDefinitionContent(id, content); err != nil {
		return err
	}

	// Save variables
	repository.DeleteVariableInfosByFlowDefID(id)
	if variables != nil {
		data, _ := json.Marshal(variables)
		var vars []model.VariableInfo
		json.Unmarshal(data, &vars)
		for i := range vars {
			vars[i].ID = 0
			vars[i].FlowDefinitionID = id
		}
		repository.BatchCreateVariableInfos(vars)
	}
	return nil
}

func DeleteFlowDefinition(id int64) error {
	repository.DeleteParametersBySource("flow", id)
	repository.DeleteVariableInfosByFlowDefID(id)
	return repository.DeleteFlowDefinition(id)
}

func GetFlowDefinitionInfo(id int64) (*FlowDefinitionInfoDTO, error) {
	fd, err := repository.GetFlowDefinitionByID(id)
	if err != nil {
		return nil, errors.New("流程定义不存在")
	}
	return buildFlowDefinitionInfoDTO(fd)
}

func PageFlowDefinitions(p *model.PageParam, flowName, flowType string) ([]FlowDefinitionInfoDTO, int64, error) {
	list, total, err := repository.PageFlowDefinitions(p, flowName, flowType)
	if err != nil {
		return nil, 0, err
	}
	dtos := make([]FlowDefinitionInfoDTO, 0, len(list))
	for _, fd := range list {
		dto, _ := buildFlowDefinitionInfoDTO(&fd)
		if dto != nil {
			dtos = append(dtos, *dto)
		}
	}
	return dtos, total, nil
}

func DeployFlowDefinition(flowDefID int64, version, remark string) error {
	fd, err := repository.GetFlowDefinitionByID(flowDefID)
	if err != nil {
		return errors.New("流程定义不存在")
	}

	// Ensure flow_info exists
	flowInfo, err := repository.GetFlowInfoByKey(fd.FlowKey)
	if err != nil {
		flowInfo = &model.FlowInfo{
			FlowKey:  fd.FlowKey,
			FlowName: fd.FlowName,
			FlowType: fd.FlowType,
			Remark:   fd.Remark,
		}
		repository.CreateFlowInfo(flowInfo)
	}

	inputs, _ := repository.ListParametersBySourceAndType("flow", fd.ID, 1)
	outputs, _ := repository.ListParametersBySourceAndType("flow", fd.ID, 2)
	variables, _ := repository.ListVariableInfosByFlowDefID(fd.ID)

	inputsJSON, _ := json.Marshal(inputs)
	outputsJSON, _ := json.Marshal(outputs)
	varsJSON, _ := json.Marshal(variables)

	fv := &model.FlowVersion{
		FlowID:            flowInfo.ID,
		FlowVersion:       version,
		FlowVersionStatus: 0,
		FlowVersionRemark: remark,
		FlowContent:       fd.FlowContent,
		Inputs:            string(inputsJSON),
		Outputs:           string(outputsJSON),
		Variables:         string(varsJSON),
	}
	return repository.CreateFlowVersion(fv)
}

func buildFlowDefinitionInfoDTO(fd *model.FlowDefinition) (*FlowDefinitionInfoDTO, error) {
	inputParams, _ := repository.ListParametersBySourceAndType("flow", fd.ID, 1)
	outputParams, _ := repository.ListParametersBySourceAndType("flow", fd.ID, 2)
	variables, _ := repository.ListVariableInfosByFlowDefID(fd.ID)

	flowContent := fd.FlowContent
	if flowContent == "" {
		flowContent = defaultFlowContent
	}

	return &FlowDefinitionInfoDTO{
		ID:               fd.ID,
		FlowKey:          fd.FlowKey,
		FlowName:         fd.FlowName,
		FlowType:         fd.FlowType,
		FlowContent:      flowContent,
		Remark:           fd.Remark,
		CreatedAt:        fd.CreatedAt,
		FlowInputParams:  inputParams,
		FlowOutputParams: outputParams,
		FlowVariables:    variables,
	}, nil
}

func saveFlowDefinitionParams(fdID int64, inputParams, outputParams interface{}) {
	repository.DeleteParametersBySource("flow", fdID)

	var allParams []model.Parameter
	if inputParams != nil {
		data, _ := json.Marshal(inputParams)
		var params []model.Parameter
		json.Unmarshal(data, &params)
		for i := range params {
			params[i].SourceType = "flow"
			params[i].SourceID = &fdID
			pt := 1
			params[i].ParamType = &pt
		}
		allParams = append(allParams, params...)
	}
	if outputParams != nil {
		data, _ := json.Marshal(outputParams)
		var params []model.Parameter
		json.Unmarshal(data, &params)
		for i := range params {
			params[i].SourceType = "flow"
			params[i].SourceID = &fdID
			pt := 2
			params[i].ParamType = &pt
		}
		allParams = append(allParams, params...)
	}
	if len(allParams) > 0 {
		repository.BatchCreateParameters(allParams)
	}
}

func generateFlowKey() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, 8)
	for i := range b {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return fmt.Sprintf("flow_%s", string(b))
}

File: server/src/main/java/com/javaoffers/nexus/core/service/FlowDefinitionService.java
package com.javaoffers.nexus.core.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.javaoffers.nexus.core.model.FlowDefinition;
import com.javaoffers.nexus.core.model.FlowInfo;
import com.javaoffers.nexus.core.model.FlowVersion;
import com.javaoffers.nexus.core.model.PageParam;
import com.javaoffers.nexus.core.model.Parameter;
import com.javaoffers.nexus.core.model.VariableInfo;
import com.javaoffers.nexus.core.repository.FlowDefinitionMapper;
import com.javaoffers.nexus.core.repository.FlowInfoMapper;
import com.javaoffers.nexus.core.repository.FlowVersionMapper;
import com.javaoffers.nexus.core.repository.ParameterMapper;
import com.javaoffers.nexus.core.repository.VariableInfoMapper;
import com.javaoffers.nexus.pkg.exception.BizException;
import lombok.Data;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/** 对应 Go core/service/flow_definition_service.go。 */
@Service
public class FlowDefinitionService {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    static final String DEFAULT_FLOW_CONTENT =
            "[{\"key\":\"start_1\",\"name\":\"开始\",\"outgoings\":[\"end_1\"],\"incomings\":[],\"elementType\":\"START\"},"
            + "{\"key\":\"end_1\",\"name\":\"结束\",\"outgoings\":[],\"incomings\":[\"start_1\"],\"elementType\":\"END\"}]";

    private final FlowDefinitionMapper flowDefinitionMapper;
    private final FlowInfoMapper flowInfoMapper;
    private final FlowVersionMapper flowVersionMapper;
    private final ParameterMapper parameterMapper;
    private final VariableInfoMapper variableInfoMapper;

    public FlowDefinitionService(FlowDefinitionMapper flowDefinitionMapper,
                                 FlowInfoMapper flowInfoMapper,
                                 FlowVersionMapper flowVersionMapper,
                                 ParameterMapper parameterMapper,
                                 VariableInfoMapper variableInfoMapper) {
        this.flowDefinitionMapper = flowDefinitionMapper;
        this.flowInfoMapper = flowInfoMapper;
        this.flowVersionMapper = flowVersionMapper;
        this.parameterMapper = parameterMapper;
        this.variableInfoMapper = variableInfoMapper;
    }

    @Data
    public static class FlowDefinitionInfoDTO {
        private Long id;
        private String flowKey;
        private String flowName;
        private String flowType;
        private String flowContent;
        private String remark;
        private Date createdAt;
        private List<Parameter> flowInputParams;
        private List<Parameter> flowOutputParams;
        private List<VariableInfo> flowVariables;
    }

    public void createFlowDefinition(String flowName, String flowType, String remark,
                                    List<Parameter> inputParams, List<Parameter> outputParams) {
        FlowDefinition fd = new FlowDefinition();
        fd.setFlowKey(generateFlowKey());
        fd.setFlowName(flowName);
        fd.setFlowType(flowType);
        fd.setFlowContent(DEFAULT_FLOW_CONTENT);
        fd.setRemark(remark);
        flowDefinitionMapper.create(fd);
        fd = flowDefinitionMapper.getByKey(fd.getFlowKey());
        saveFlowDefinitionParams(fd.getId(), inputParams, outputParams);
    }

    public void updateFlowDefinition(Long id, String flowName, String flowType, String remark,
                                     List<Parameter> inputParams, List<Parameter> outputParams) {
        FlowDefinition fd = new FlowDefinition();
        fd.setId(id);
        fd.setFlowName(flowName);
        fd.setFlowType(flowType);
        fd.setRemark(remark);
        flowDefinitionMapper.update(fd);
        saveFlowDefinitionParams(id, inputParams, outputParams);
    }

    public void saveFlowDefinitionContent(Long id, String content, List<VariableInfo> variables) {
        flowDefinitionMapper.updateContent(id, content);
        variableInfoMapper.deleteByFlowDefId(id);
        if (variables != null) {
            for (VariableInfo v : variables) {
                v.setId(null);
                v.setFlowDefinitionId(id);
            }
            variableInfoMapper.batchCreate(variables);
        }
    }

    public void deleteFlowDefinition(Long id) {
        parameterMapper.deleteBySource("flow", id);
        variableInfoMapper.deleteByFlowDefId(id);
        flowDefinitionMapper.delete(id);
    }

    public FlowDefinitionInfoDTO getFlowDefinitionInfo(Long id) {
        FlowDefinition fd = flowDefinitionMapper.getById(id);
        if (fd == null) {
            throw new BizException("流程定义不存在");
        }
        return buildDTO(fd);
    }

    public Page<FlowDefinitionInfoDTO> pageFlowDefinitions(PageParam p, String flowName, String flowType) {
        int pn = p.getPageNum() <= 0 ? 1 : p.getPageNum();
        int ps = p.getPageSize() <= 0 ? 10 : p.getPageSize();
        List<FlowDefinition> list = flowDefinitionMapper.page(pn, ps, flowName, flowType);
        long total = flowDefinitionMapper.count(flowName, flowType);
        List<FlowDefinitionInfoDTO> dtos = new ArrayList<>();
        for (FlowDefinition fd : list) {
            dtos.add(buildDTO(fd));
        }
        return new Page<>(total, dtos);
    }

    public void deployFlowDefinition(Long flowDefID, String version, String remark) {
        FlowDefinition fd = flowDefinitionMapper.getById(flowDefID);
        if (fd == null) {
            throw new BizException("流程定义不存在");
        }

        FlowInfo flowInfo = flowInfoMapper.getByKey(fd.getFlowKey());
        if (flowInfo == null) {
            flowInfo = new FlowInfo();
            flowInfo.setFlowKey(fd.getFlowKey());
            flowInfo.setFlowName(fd.getFlowName());
            flowInfo.setFlowType(fd.getFlowType());
            flowInfo.setRemark(fd.getRemark());
            flowInfoMapper.create(flowInfo);
            flowInfo = flowInfoMapper.getByKey(fd.getFlowKey());
        }

        List<Parameter> inputs = parameterMapper.listBySourceAndType("flow", fd.getId(), 1);
        List<Parameter> outputs = parameterMapper.listBySourceAndType("flow", fd.getId(), 2);
        List<VariableInfo> variables = variableInfoMapper.listByFlowDefId(fd.getId());

        FlowVersion fv = new FlowVersion();
        fv.setFlowId(flowInfo.getId());
        fv.setFlowVersion(version);
        fv.setFlowVersionStatus(0);
        fv.setFlowVersionRemark(remark);
        fv.setFlowContent(fd.getFlowContent());
        fv.setInputs(toJson(inputs));
        fv.setOutputs(toJson(outputs));
        fv.setVariables(toJson(variables));
        flowVersionMapper.create(fv);
    }

    private FlowDefinitionInfoDTO buildDTO(FlowDefinition fd) {
        FlowDefinitionInfoDTO dto = new FlowDefinitionInfoDTO();
        dto.setId(fd.getId());
        dto.setFlowKey(fd.getFlowKey());
        dto.setFlowName(fd.getFlowName());
        dto.setFlowType(fd.getFlowType());
        String content = fd.getFlowContent();
        dto.setFlowContent((content == null || content.isEmpty()) ? DEFAULT_FLOW_CONTENT : content);
        dto.setRemark(fd.getRemark());
        dto.setCreatedAt(fd.getCreatedAt());
        dto.setFlowInputParams(parameterMapper.listBySourceAndType("flow", fd.getId(), 1));
        dto.setFlowOutputParams(parameterMapper.listBySourceAndType("flow", fd.getId(), 2));
        dto.setFlowVariables(variableInfoMapper.listByFlowDefId(fd.getId()));
        return dto;
    }

    private void saveFlowDefinitionParams(Long fdID, List<Parameter> inputParams, List<Parameter> outputParams) {
        parameterMapper.deleteBySource("flow", fdID);
        List<Parameter> all = new ArrayList<>();
        if (inputParams != null) {
            for (Parameter p : inputParams) {
                p.setSourceType("flow");
                p.setSourceId(fdID);
                p.setParamType(1);
            }
            all.addAll(inputParams);
        }
        if (outputParams != null) {
            for (Parameter p : outputParams) {
                p.setSourceType("flow");
                p.setSourceId(fdID);
                p.setParamType(2);
            }
            all.addAll(outputParams);
        }
        parameterMapper.batchCreate(all);
    }

    private String generateFlowKey() {
        String chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        StringBuilder sb = new StringBuilder("flow_");
        for (int i = 0; i < 8; i++) {
            sb.append(chars.charAt((int) (Math.random() * chars.length())));
        }
        return sb.toString();
    }

    private String toJson(Object obj) {
        try {
            return MAPPER.writeValueAsString(obj);
        } catch (Exception e) {
            return "[]";
        }
    }
}
